sys/drv/sd.c

```diff
@@ -0,0 +1,371 @@
+ */
+int sd_write_blk(device_t *dev, uint32_t lba, void *buffer, uint32_t num) {
+  assert(dev->driver == (driver_t *)&sd_block_device_driver);
+  sd_state_t *state = (sd_state_t *)dev->state;
+
+  uint32_t *buf = (uint32_t *)buffer;
+  int r;
+  uint32_t c = 0;
+
+  if (num < 1)
+    return 0;
+
+  if (state->props & SD_SUPP_BLKCNT) {
+    if (num > 1 && (state->props & SD_SUPP_BLKCNT)) {
+      r = emmc_send_cmd(dev, EMMC_CMD(SET_BLOCK_COUNT), num, NULL);
+      if (r)
+        return 0;
+    }
+    emmc_set_prop(dev, EMMC_PROP_RW_BLKCNT, num);
+    emmc_set_prop(dev, EMMC_PROP_RW_BLKSIZE, DEFAULT_BLKSIZE);
+
+    emmc_send_cmd(
+      dev, num == 1 ? EMMC_CMD(WRITE_BLOCK) : EMMC_CMD(WRITE_MULTIPLE_BLOCKS),
+      lba, NULL);
+    if (emmc_wait(dev, EMMC_I_WRITE_READY))
+      return 0;
+  } else {
+    emmc_set_prop(dev, EMMC_PROP_RW_BLKCNT, num);
+    emmc_set_prop(dev, EMMC_PROP_RW_BLKSIZE, DEFAULT_BLKSIZE);
+  }
+  while (c < num) {
+    if (!(state->props & SD_SUPP_BLKCNT)) {
+      emmc_send_cmd(dev, EMMC_CMD(WRITE_BLOCK), lba + c, NULL);
+      if (emmc_wait(dev, EMMC_I_WRITE_READY))
+        return 0;
+    }
+    if (emmc_write(dev, buf, DEFAULT_BLKSIZE, NULL)) {
+      klog("e.MMC write failed");
+      return 0;
+    }
+    if (emmc_wait(dev, EMMC_I_DATA_DONE))
+      return 0;
+
+    c++;
+    buf += 128;
+  }
+
+  return num * DEFAULT_BLKSIZE;
+}
+
+static inline int sd_check_uio(uio_t *uio) {
+  if (uio->uio_resid & (DEFAULT_BLKSIZE - 1))
+    return -1;
+  if (uio->uio_offset & (DEFAULT_BLKSIZE - 1))
+    return -1;
+  return 0;
+}
+
+typedef int (*sd_transfer_t)(device_t *dev, uint32_t lba, void *buffer,
+                             uint32_t num);
+
+static int sd_dop_uio(devnode_t *d, uio_t *uio) {
+  device_t *dev = d->data;
+  sd_state_t *state = (sd_state_t *)dev->state;
+
+  if (sd_check_uio(uio))
+    return -1;
+
+  uint32_t lba = uio->uio_offset / DEFAULT_BLKSIZE;
+  uint32_t blk_cnt = uio->uio_resid / DEFAULT_BLKSIZE;
+
+  sd_transfer_t transfer = uio->uio_op == UIO_READ ? sd_read_blk : sd_write_blk;
+
+  for (uint32_t blocks_left = blk_cnt; blocks_left;
+       blocks_left -= SD_KERNEL_BLOCKS) {
+    uint32_t num = min(blocks_left, (uint32_t)SD_KERNEL_BLOCKS);
+    transfer(dev, lba, state->block_buf, num);
+    uiomove_frombuf(state->block_buf, num * DEFAULT_BLKSIZE, uio);
+    if (num < SD_KERNEL_BLOCKS)
+      break;
+  }
+
+  return 0;
+}
+
+static devops_t sd_devops = {
+  .d_read = sd_dop_uio,
+  .d_write = sd_dop_uio,
+};
+
+static int sd_attach(device_t *dev) {
+  sd_state_t *state = (sd_state_t *)dev->state;
+
+  state->block_buf = kmalloc(M_DEV, SD_KERNEL_BLOCKS * DEFAULT_BLKSIZE, 0);
+
+  if (sd_init(dev))
+    return -1;
+
+  devfs_makedev_new(NULL, "sd_card", &sd_devops, dev, NULL);
+
+  return 0;
+}
+
+static driver_t sd_block_device_driver = {
+  .desc = "SD(SC/HC) block device driver",
+  .size = sizeof(sd_state_t),
+  .probe = sd_probe,
+  .attach = sd_attach,
+  .pass = SECOND_PASS,
```

Comment:
`pass` should be placed immediately under the `size` (this is a convention suggested by @cahirwpz).

Reply:
Fixed.